lua/wire/zvm/zvm_tests.lua

```diff
@@ -0,0 +1,411 @@
+--------------------------------------------------------------------------------
+-- Library for making and running automated tests for Zyelios Virtual Machine Programs
+--
+--
+--------------------------------------------------------------------------------
+
+TESTING = true
+include("wire/cpulib.lua")
+include("wire/client/hlzasm/hc_compiler.lua")
+
+
+ZVMTestSuite = {
+	TestFiles = {},
+	TestQueue = {},
+	TestStatuses = {},
+	Warnings = 0,
+	CurrentWarnings = 0
+}
+
+local testDirectory = "wire/zvm/tests"
+
+function ZVMTestSuite.CMDRun(_,_,_,names)
+	ZVMTestSuite.Warnings = 0
+	ZVMTestSuite.TestFiles = {}
+	for filename in string.gmatch(names, '[^,]+') do
+		local files = file.Find('lua/'..testDirectory..'/'..filename..'.lua',"GAME")
+			for _,i in ipairs(files) do
+				ZVMTestSuite.TestFiles[#ZVMTestSuite.TestFiles+1] = i
+			end
+	end
+	if #ZVMTestSuite.TestFiles == 0 then
+		if names ~= nil then
+			if names ~= "" then
+				print("Didn't find any tests with name(s): "..names)
+				return
+			end
+		end
+		ZVMTestSuite.RunAll()
+	else
+		PrintTable(ZVMTestSuite.TestFiles)
+		ZVMTestSuite.StartTesting()
+	end
+end
+
+function ZVMTestSuite.RunAll()
+	local files,directories = file.Find(testDirectory..'/*.lua',"LUA","nameasc")
+	ZVMTestSuite.TestFiles = files or {}
+	ZVMTestSuite.StartTesting()
+end
+
+function ZVMTestSuite.StartTesting()
+	ZVMTestSuite.TestQueue = {}
+	ZVMTestSuite.TestStatuses = {}
+	for ind,i in ipairs(ZVMTestSuite.TestFiles) do -- copy with reversed indexes so we can use cheap popping
+		ZVMTestSuite.TestQueue[(#ZVMTestSuite.TestFiles)+1-ind] = i
+	end
+	print(#ZVMTestSuite.TestFiles.." tests loaded")
+	ZVMTestSuite.RunNextTest()
+end
+
+function ZVMTestSuite.FinishTest(fail)
+	local finalFail = false
+	if fail == nil then
+		finalFail = true
+	else
+		finalFail = fail
+	end
+	if ZVMTestSuite.CurrentWarnings > 0 then
+		print("Compiler Warnings from "..ZVMTestSuite.TestQueue[#ZVMTestSuite.TestQueue]..": "..ZVMTestSuite.CurrentWarnings)
```

Comment:
I don't think showing warnings is necessary.

Reply:
My intent was that they can be used to try and catch regressions in compilation & syntax, if we checked a prior commit for how many warns were produced vs now we can try to track down code changes that cause more warnings to determine if the warnings are intentional or not.

Though this will only display the NUMBER of warnings the test had, and not any of the actual warning text itself for the moment.

Right now all the warnings that are in during the state of this PR include:
["Trying to generate invalid code"](https://github.com/wiremod/wire-cpu/blob/ac69f0228d8b7d795e6f1778d22ae2857c486b1b/lua/wire/client/hlzasm/hc_codetree.lua#L291C1-L291C1)
["Instruction "(opcode name)" is obsolete/an old mnemonic"](https://github.com/wiremod/wire-cpu/blob/ac69f0228d8b7d795e6f1778d22ae2857c486b1b/lua/wire/client/hlzasm/hc_syntax.lua#L41)
["Using complex expression as operand: might corrupt user register"](https://github.com/wiremod/wire-cpu/blob/ac69f0228d8b7d795e6f1778d22ae2857c486b1b/lua/wire/client/hlzasm/hc_syntax.lua#L97C1-L97C89)
["Warning: using an unpreserved register"](https://github.com/wiremod/wire-cpu/blob/ac69f0228d8b7d795e6f1778d22ae2857c486b1b/lua/wire/client/hlzasm/hc_syntax.lua#L195C1-L195C61)
["Defining a vector inside a function block might cause issues"](https://github.com/wiremod/wire-cpu/blob/ac69f0228d8b7d795e6f1778d22ae2857c486b1b/lua/wire/client/hlzasm/hc_syntax.lua#L925)

Which, I agree that a lot of these aren't exactly necessary to test for, but "Trying to generate invalid code" indicates labels and registers not being used in expressions, which could occur if a user expression (ZAP R0 R1 not parsing R1 for example should generate this err) exits before all of the values are used.